ethcore/src/state_db.rs

```diff
@@ -0,0 +1,160 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use lru_cache::LruCache;
+use util::journaldb::JournalDB;
+use util::hash::{H256};
+use util::hashdb::HashDB;
+use util::{Arc, Address, DBTransaction, UtilError, Mutex};
+use account::Account;
+
+const STATE_CACHE_ITEMS: usize = 65536;
+
+struct AccountCache {
+	accounts: LruCache<Address, Option<Account>>,
+}
+
+/// State database abstraction.
+/// Manages shared global state cache.
+/// A clone of `StateDB` may be created as canonical or not.
+/// For canonical clones cache changes are accumulated and applied
+/// on commit.
+/// For non-canonical clones cache is cleared on commit.
+pub struct StateDB {
+	db: Box<JournalDB>,
+	account_cache: Arc<Mutex<AccountCache>>,
+	cache_overlay: Vec<(Address, Option<Account>)>,
+	is_canon: bool,
+}
+
+impl StateDB {
+	/// Create a new instance wrapping `JournalDB`
+	pub fn new(db: Box<JournalDB>) -> StateDB {
+		StateDB {
+			db: db,
+			account_cache: Arc::new(Mutex::new(AccountCache { accounts: LruCache::new(STATE_CACHE_ITEMS) })),
+			cache_overlay: Vec::new(),
+			is_canon: false,
+		}
+	}
+
+	/// Commit all recent insert operations and canonical historical commits' removals from the
+	/// old era to the backing database, reverting any non-canonical historical commit's inserts.
+	pub fn commit(&mut self, batch: &DBTransaction, now: u64, id: &H256, end: Option<(u64, H256)>) -> Result<u32, UtilError> {
+		let records = try!(self.db.commit(batch, now, id, end));
+		if self.is_canon {
+			self.commit_cache();
+		} else {
+			self.clear_cache();
+		}
+		Ok(records)
+	}
+
+	/// Returns an interface to HashDB.
+	pub fn as_hashdb(&self) -> &HashDB {
+		self.db.as_hashdb()
+	}
+
+	/// Returns an interface to mutable HashDB.
+	pub fn as_hashdb_mut(&mut self) -> &mut HashDB {
+		self.db.as_hashdb_mut()
+	}
+
+	/// Clone the database.
+	pub fn boxed_clone(&self) -> StateDB {
+		StateDB {
+			db: self.db.boxed_clone(),
+			account_cache: self.account_cache.clone(),
+			cache_overlay: Vec::new(),
+			is_canon: false,
+		}
+	}
+
+	/// Clone the database for a canonical state.
+	pub fn boxed_clone_canon(&self) -> StateDB {
+		StateDB {
+			db: self.db.boxed_clone(),
+			account_cache: self.account_cache.clone(),
+			cache_overlay: Vec::new(),
+			is_canon: true,
+		}
+	}
+
+	/// Check if pruning is enabled on the database.
+	pub fn is_pruned(&self) -> bool {
+		self.db.is_pruned()
+	}
+
+	/// Heap size used.
+	pub fn mem_used(&self) -> usize {
+		self.db.mem_used() //TODO: + self.account_cache.lock().heap_size_of_children()
+	}
+
+	/// Returns underlying `JournalDB`.
+	pub fn journal_db(&self) -> &JournalDB {
+		&*self.db
+	}
+
+	/// Enqueue cache change.
+	pub fn cache_account(&mut self, addr: Address, data: Option<Account>) {
+		self.cache_overlay.push((addr, data));
+	}
+
+	/// Apply pending cache changes.
+	fn commit_cache(&mut self) {
```

Comment:
this doesn't write through to the DB - is that not an issue on unexpected shutdown?